presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControl.java

```diff
@@ -0,0 +1,328 @@
+import static com.facebook.presto.spi.security.AccessDeniedException.denyShowTables;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static java.util.Objects.requireNonNull;
+
+public class FileBasedSystemAccessControl
+        implements SystemAccessControl
+{
+    private static final String CATALOG_ACCESS_DENIED = "Access denied for catalog ";
+    public static final String NAME = "file";
+
+    private static final FileBasedSystemAccessControl INSTANCE = new FileBasedSystemAccessControl(jsonCodec(FileBasedSystemAccessControlRules.class));
+
+    private final JsonCodec<FileBasedSystemAccessControlRules> rulesJsonCodec;
+    @JsonProperty("catalogs") private List<CatalogAccessControlRule> catalogRules;
+
+    @Inject
+    private FileBasedSystemAccessControl(JsonCodec<FileBasedSystemAccessControlRules> rulesCodec)
+    {
+        rulesJsonCodec = rulesCodec;
+    }
+
+    public static class Factory
+            implements SystemAccessControlFactory
+    {
+        @Override
+        public String getName()
+        {
+            return NAME;
+        }
+
+        @Override
+        public SystemAccessControl create(Map<String, String> config)
+        {
+            final String CONFIG_FILE_NAME = "security.config-file";
+            requireNonNull(config, "config is null");
+
+            try {
+                Path path = Paths.get(config.get(CONFIG_FILE_NAME));
+                if (!path.isAbsolute()) {
+                    path = path.toAbsolutePath();
+                }
+                FileBasedSystemAccessControlRules rules = INSTANCE.rulesJsonCodec.fromJson(Files.readAllBytes(path));
+                INSTANCE.catalogRules = rules.getCatalogRules();
+            }
+            catch (NullPointerException e) {
+                throw new ConfigurationNotFoundException(CONFIG_FILE_NAME, "while loading file based system access control");
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+            return INSTANCE;
+        }
+    }
+
+    @Override
+    public void checkCanSetUser(Principal principal, String userName)
+    {
+    }
+
+    @Override
+    public void checkCanSetSystemSessionProperty(Identity identity, String propertyName)
+    {
+    }
+
+    @Override
+    public Set<String> filterCatalogs(Identity identity, Set<String> catalogs)
+    {
+        ImmutableSet.Builder<String> filteredCatalogs = ImmutableSet.builder();
+        for (String catalog : catalogs) {
+            if (canAccessCatalog(identity, catalog)) {
+                filteredCatalogs.add(catalog);
+            }
+        }
+        return filteredCatalogs.build();
+    }
+
+    @Override
+    public void checkCanCreateSchema(Identity identity, CatalogSchemaName schema)
+    {
+        if (!canAccessCatalog(identity, schema.getCatalogName())) {
+            denyCreateSchema(schema.toString(), CATALOG_ACCESS_DENIED + schema.getCatalogName());
```

Comment:
I'm also not sure you need to check canAccessCatalog in all of these methods, because you wouldn't get as far as canCreateSchema if you've filtered out the catalog.